smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,358 @@
+            sim_name (Optional[str], optional): Simulation name prefix. Defaults to None.
+            seed (int, optional): Seed for the first environment. Defaults to 42.
+
+        Raises:
+            TypeError: If any environment constructor is not callable.
+        """
+
+        if len(env_constructors) > mp.cpu_count():
+            warnings.warn(
+                f"Simulation might slow down, since the requested number of parallel "
+                f"environments ({len(env_constructors)}) exceed the number of available "
+                f"CPU cores ({mp.cpu_count()}).",
+                ResourceWarning,
+            )
+
+        if any([not callable(ctor) for ctor in env_constructors]):
+            raise TypeError(
+                f"Found non-callable `env_constructors`. Expected `env_constructors` of type "
+                f"`Sequence[Callable[[], gym.Env]]`, but got {env_constructors})."
+            )
+
+        # Worker polling period in seconds.
+        self._polling_period = 0.1
+
+        # Fork is not a thread safe method.
+        forkserver_available = "forkserver" in mp.get_all_start_methods()
+        start_method = "forkserver" if forkserver_available else "spawn"
+        mp_ctx = mp.get_context(start_method)
+
+        self.env_constructors = env_constructors
+
+        self.error_queue = mp_ctx.Queue()
+        self.parent_pipes = []
+        self.processes = []
+        for idx, env_constructor in enumerate(self.env_constructors):
+            parent_pipe, child_pipe = mp_ctx.Pipe()
+            process = mp_ctx.Process(
+                target=_worker,
+                name=f"Worker<{type(self).__name__}>-<{idx}>",
+                args=(
+                    idx,
+                    sim_name,
+                    CloudpickleWrapper(env_constructor),
+                    auto_reset,
+                    child_pipe,
+                    self.error_queue,
+                    self._polling_period,
+                ),
+            )
+            self.parent_pipes.append(parent_pipe)
+            self.processes.append(process)
+
+            # Daemonic subprocesses quit when parent process quits. However, daemonic
+            # processes cannot spawn children. Hence, `process.daemon` is set to False.
+            process.daemon = False
+            process.start()
+            child_pipe.close()
+
+        self._state = AsyncState.DEFAULT
+
+        # Wait for all environments to successfully startup
+        _, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+
+        # Get and check observation and action spaces
+        observation_space, action_space = self._get_spaces()
+
+        # TODO: This dummy `observation_space` and `action_space` should be removed after they
+        # are properly specified in SMARTS/smarts/env/hiway_env.py:__init__() function.
```

Comment:
Please reference an issue.

Reply:
Now, `observation_space` and `action_space` inside `ParallelEnv` class are obtained from the single environment.